libraries/chain_kv/include/b1/session/rocks_data_store.hpp

```diff
@@ -0,0 +1,684 @@
+   rocksdb::WriteOptions                        m_write_options;
+};
+
+template <typename Allocator>
+rocks_data_store<Allocator> make_rocks_data_store(std::shared_ptr<rocksdb::DB> db,
+                                                  std::shared_ptr<Allocator>   memory_allocator) {
+   return { std::move(db), std::move(memory_allocator) };
+}
+
+template <typename Allocator>
+rocks_data_store<Allocator>::rocks_data_store(std::shared_ptr<rocksdb::DB> db)
+    : m_db{ std::move(db) }, m_allocator{ std::make_shared<Allocator>() } {}
+
+template <typename Allocator>
+rocks_data_store<Allocator>::rocks_data_store(std::shared_ptr<rocksdb::DB> db,
+                                              std::shared_ptr<Allocator>   memory_allocator)
+    : m_db{ std::move(db) }, m_allocator{ std::move(memory_allocator) } {}
+
+template <typename Allocator>
+const key_value rocks_data_store<Allocator>::read(const shared_bytes& key) const {
+   if (!m_db) {
+      return key_value::invalid;
+   }
+
+   auto key_slice      = rocksdb::Slice{ reinterpret_cast<const char*>(key.data()), key.length() };
+   auto pinnable_value = rocksdb::PinnableSlice{};
+   auto status         = m_db->Get(m_read_options, column_family_(), key_slice, &pinnable_value);
+
+   if (status.code() != rocksdb::Status::Code::kOk) {
+      return key_value::invalid;
+   }
+
+   return make_kv(key_slice.data(), key_slice.size(), pinnable_value.data(), pinnable_value.size(), m_allocator);
+}
+
+template <typename Allocator>
+void rocks_data_store<Allocator>::write(key_value kv) {
+   if (!m_db) {
+      return;
+   }
+
+   auto key_slice   = rocksdb::Slice{ reinterpret_cast<const char*>(kv.key().data()), kv.key().length() };
+   auto value_slice = rocksdb::Slice{ reinterpret_cast<const char*>(kv.value().data()), kv.value().length() };
+   auto status      = m_db->Put(m_write_options, column_family_(), key_slice, value_slice);
+}
+
+template <typename Allocator>
+bool rocks_data_store<Allocator>::contains(const shared_bytes& key) const {
+   if (!m_db) {
+      return false;
+   }
+
+   auto key_slice = rocksdb::Slice{ reinterpret_cast<const char*>(key.data()), key.length() };
+   auto value     = std::string{};
+   return m_db->KeyMayExist(m_read_options, column_family_(), key_slice, &value);
+}
+
+template <typename Allocator>
+void rocks_data_store<Allocator>::erase(const shared_bytes& key) {
+   if (!m_db) {
+      return;
+   }
+
+   auto key_slice = rocksdb::Slice{ reinterpret_cast<const char*>(key.data()), key.length() };
+   auto status    = m_db->Delete(m_write_options, column_family_(), key_slice);
+}
+
+template <typename Allocator>
+void rocks_data_store<Allocator>::clear() {
+   // TODO:
+}
+
+template <typename Allocator>
+template <typename Iterable, typename Other_allocator>
+const std::pair<std::vector<key_value>, std::unordered_set<shared_bytes>>
+rocks_data_store<Allocator>::read_(const Iterable& keys, const std::shared_ptr<Other_allocator>& a) const {
```

Comment:
I would argue that the type names chosen here are quality.  Just because they are long doesn't lessen the quality.  If the coding style guidelines details how non public members of a type should be named then I can change them to follow that.